src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncContext.Unix.cs

```diff
@@ -1480,7 +1480,8 @@ public SocketError Connect(byte[] socketAddress, int socketAddressLen)
             SocketError errorCode;
             int observedSequenceNumber;
             _sendQueue.IsReady(this, out observedSequenceNumber);
-            if (SocketPal.TryStartConnect(_socket, socketAddress, socketAddressLen, out errorCode))
+            if (SocketPal.TryStartConnect(_socket, socketAddress, socketAddressLen, out errorCode) ||
+                !ShouldRetrySyncOperation(out errorCode))
```

Comment:
I have trouble understanding this
`TryStartXXX || !ShouldRetrySyncOperation(out errorCode)` pattern.

The `TryStart` method returns `false` when `EAGAIN`. `EAGAIN` can only happen when the socket is non-blocking. When `_nonBlockingSet` is `true`, `ShouldRetrySyncOperation` returns `true`.

`false || !true` is still `false`.

Reply:
See here: https://github.com/dotnet/runtime/blob/main/src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncContext.Unix.cs#L1398

Apparently we get EAGAIN for timeouts on blocking sockets.

I merged this fix into my request retry PR, which was consistently seeing the SocketAsyncContext assert, and the assert doesn't seem to repro anymore. That seems like pretty good evidence that this fix is correct.

Thoughts?

Reply:
Ah, yes, for timeouts it can happen as well.

From man 7 socket:

> if  no  data  has been  transferred  and  the timeout has been reached, then -1 is returned with errno set to EAGAIN or EWOULDBLOCK, or EINPROGRESS (for connect(2)) just as if the socket was specified to  be  nonblocking